client/rpc/src/state/tests.rs

```diff
@@ -233,113 +213,107 @@ fn should_return_child_storage_entries() {
 	let client = Arc::new(substrate_test_runtime_client::new());
 	let genesis_hash = client.genesis_hash();
 	let (client, _child) =
-		new_full(client, SubscriptionManager::new(Arc::new(TaskExecutor)), DenyUnsafe::No, None);
+		new_full(client, SubscriptionTaskExecutor::new(TaskExecutor), DenyUnsafe::No, None);
+
+	use jsonrpsee::types::{CallError, Error};
 
 	assert_matches!(
-		executor::block_on(client.call(
-			"balanceOf".into(),
-			Bytes(vec![1, 2, 3]),
-			Some(genesis_hash).into()
-		)),
-		Err(Error::Client(_))
+		client
+			.call("balanceOf".into(), Bytes(vec![1, 2, 3]), Some(genesis_hash).into())
+			.await,
+		Err(Error::Call(CallError::Failed(_)))
 	)
 }
 
-#[test]
-fn should_notify_about_storage_changes() {
-	let (subscriber, id, mut transport) = Subscriber::new_test("test");
-
-	{
-		let mut client = Arc::new(substrate_test_runtime_client::new());
-		let (api, _child) = new_full(
-			client.clone(),
-			SubscriptionManager::new(Arc::new(TaskExecutor)),
-			DenyUnsafe::No,
-			None,
-		);
-
-		api.subscribe_storage(Default::default(), subscriber, None.into());
-
-		// assert id assigned
-		assert!(matches!(executor::block_on(id), Ok(Ok(SubscriptionId::String(_)))));
-
-		let mut builder = client.new_block(Default::default()).unwrap();
-		builder
-			.push_transfer(runtime::Transfer {
-				from: AccountKeyring::Alice.into(),
-				to: AccountKeyring::Ferdie.into(),
-				amount: 42,
-				nonce: 0,
-			})
-			.unwrap();
-		let block = builder.build().unwrap().block;
-		executor::block_on(client.import(BlockOrigin::Own, block)).unwrap();
-	}
-
-	// Check notification sent to transport
-	executor::block_on((&mut transport).take(2).collect::<Vec<_>>());
-	assert!(executor::block_on(transport.next()).is_none());
+#[tokio::test]
+async fn should_notify_about_storage_changes() {
+	let mut client = Arc::new(substrate_test_runtime_client::new());
+	let (api, _child) =
+		new_full(client.clone(), SubscriptionTaskExecutor::new(TaskExecutor), DenyUnsafe::No, None);
+
+	let api_rpc = api.into_rpc();
+	let (_sub_id, mut sub_rx) = api_rpc.test_subscription("state_subscribeStorage", None).await;
+
+	// Cause a change:
+	let mut builder = client.new_block(Default::default()).unwrap();
+	builder
+		.push_transfer(runtime::Transfer {
+			from: AccountKeyring::Alice.into(),
+			to: AccountKeyring::Ferdie.into(),
+			amount: 42,
+			nonce: 0,
+		})
+		.unwrap();
+	let block = builder.build().unwrap().block;
+	client.import(BlockOrigin::Own, block).await.unwrap();
+
+	// We should get a message back on our subscription about the storage change:
+	// TODO (jsdw): previously we got back 2 messages here.
+	// TODO (dp): I agree that we differ here. I think `master` always includes the initial value of
+	// the storage?
+	let msg = timeout_secs(5, sub_rx.next()).await;
+	assert_matches!(&msg, Ok(Some(json)) => {
+		serde_json::from_str::<SubscriptionResponse<StorageChangeSet<H256>>>(&json).expect("The right kind of response")
+	});
```

Comment:
@niklasad1 ^^^

Reply:
Yeah, master sends back `empty changes` and we don't, that includes the initial value so that makes sense to me.

Reply:
I think ours is the better behaviour, but yeah, I hope nobody relied on the old behaviour. I guess polkadot.js has code to filter out empties.

Reply:
https://github.com/paritytech/substrate/blob/master/client/rpc/src/state/state_full.rs#L544-#L549
https://github.com/paritytech/substrate/blob/master/client/rpc/src/state/state_full.rs#L560-#L563

this might be empty :point_up: 

our branch:
https://github.com/paritytech/substrate/blob/dp-jsonrpsee-integration-2/client/rpc/src/state/state_full.rs#L523-#L526
https://github.com/paritytech/substrate/blob/dp-jsonrpsee-integration-2/client/rpc/src/state/state_full.rs#L544-#L548